fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -664,6 +684,210 @@ public void exec() throws Exception {
+                if (v != null) {
+                    inDegrees.put(child, v + 1);
+                } else {
+                    inDegrees.put(child, 1);
+                    queue.add(child);
+                }
+            }
+        }
+
+        if (fragments.size() != inDegrees.size()) {
+            for (PlanFragment fragment : fragments) {
+                if (!inDegrees.containsKey(fragment)) {
+                    LOG.warn("This fragment does not belong to the fragment tree: {}", fragment.getFragmentId());
+                }
+            }
+            throw new StarRocksPlannerException("Some fragments do not belong to the fragment tree",
+                    ErrorType.INTERNAL_ERROR);
+        }
+
+        // `queue` contains the fragments whose in-degree is zero.
+        queue.add(root);
+        List<List<PlanFragment>> results = Lists.newArrayList();
+        int numOutputFragments = 0;
+        while (!queue.isEmpty()) {
+            int size = queue.size();
+            numOutputFragments += size;
+            List<PlanFragment> curResults = new ArrayList<>(size);
+            // The next `size` fragments can be delivered concurrently, because zero in-degree indicates that
+            // they don't depend on each other and all the fragments depending on them have been delivered.
+            for (int i = 0; i < size; ++i) {
+                PlanFragment fragment = queue.poll();
+                curResults.add(fragment);
+
+                for (PlanFragment child : fragment.getChildren()) {
+                    int degree = inDegrees.compute(child, (k, v) -> v - 1);
+                    if (degree == 0) {
+                        queue.add(child);
+                    }
+                }
+            }
+            results.add(curResults);
+        }
+        Preconditions.checkState(fragments.size() == numOutputFragments);
+
+        return results;
+    }
+
+    /**
+     * Deliver multiple fragments concurrently according to the topological order,
+     * and all the instances of a fragment to the same destination host are delivered in the same request.
+     */
+    private void deliverExecBatchFragmentsRequests() throws Exception {
+        long queryDeliveryTimeoutMs = Math.min(queryOptions.query_timeout, queryOptions.query_delivery_timeout) * 1000L;
+        List<List<PlanFragment>> topologicalFragments = computeTopologicalOrderFragments();
+
+        lock();
+        try {
+            // execute all instances from up to bottom
+            int backendNum = 0;
+            int profileFragmentId = 0;
+
+            Set<Long> dbIds = connectContext != null ? connectContext.getCurrentSqlDbIds() : null;
+            // The cached desc table is only used for the non-first instance request to the same host.
+            // Since all the instances of a fragment will be sent in a request,
+            // there is no chance to use the cached desc table.
+            descTable.setIs_cached(false);
```

Comment:
turn off the desc_tbl cache for batch delivery?
Just use a Set<TNetworkAddress> delivered_hosts to records the BE who has already received the fragments.
1. for the first time, the delivery carries desc_tbl;
2. for the second time, the delivery cast set desc_tbl cache flag to be true.

Reply:
All the instance delivered to a host is sent in **ONE** request, so there isn't second time.